### ISSUE TITLE
When I killed my java etcd program, I found that etcd would stop working and report the following error. What is the reason for this?

### ISSUE THREAD

**Issue by luomengY** [etcd-io/etcd#14314]
### What happened?

{"level":"warn","ts":"2022-08-05T14:20:13.984+0800","caller":"v3rpc/interceptor.go:197","msg":"request stats","start time":"2022-08-05T14:20:10.778+0800","time spent":"3.20534961s","remote":"10.3.71.106:49706","response type":"/v3lockpb.Lock/Lock","request count":-1,"request size":-1,"response count":-1,"response size":-1,"request content":""}}

{"level":"fatal","ts":"2022-08-05T14:31:47.331+0800","caller":"backend/batch_tx.go:152","msg":"failed to find a bucket","bucket-name":"key","stack":"go.etcd.io/etcd/server/v3/mvcc/backend.(*batchTx).unsafePut\n\t/go/src/go.etcd.io/etcd/release/etcd/server/mvcc/backend/batch_tx.go:155\ngo.etcd.io/etcd/server/v3/mvcc/backend.(*batchTx).UnsafeSeqPut\n\t/go/src/go.etcd.io/etcd/release/etcd/server/mvcc/backend/batch_tx.go:146\ngo.etcd.io/etcd/server/v3/mvcc/backend.(*batchTxBuffered).UnsafeSeqPut\n\t/go/src/go.etcd.io/etcd/release/etcd/server/mvcc/backend/batch_tx.go:368\ngo.etcd.io/etcd/server/v3/mvcc.(*storeTxnWrite).delete\n\t/go/src/go.etcd.io/etcd/release/etcd/server/mvcc/kvstore_txn.go:279\ngo.etcd.io/etcd/server/v3/mvcc.(*storeTxnWrite).deleteRange\n\t/go/src/go.etcd.io/etcd/release/etcd/server/mvcc/kvstore_txn.go:257\ngo.etcd.io/etcd/server/v3/mvcc.(*storeTxnWrite).DeleteRange\n\t/go/src/go.etcd.io/etcd/release/etcd/server/mvcc/kvstore_txn.go:102\ngo.etcd.io/etcd/server/v3/mvcc.(*metricsTxnWrite).DeleteRange\n\t/go/src/go.etcd.io/etcd/release/etcd/server/mvcc/metrics_txn.go:46\ngo.etcd.io/etcd/server/v3/lease.(*lessor).Revoke\n\t/go/src...........

### What did you expect to happen?

etcd cannot stop working when the jetcd client disconnects unexpectedly.

### How can we reproduce it (as minimally and precisely as possible)?

Connect etcd through java etcd client api, then store key-value and read key-value in etcd, and manually kill the java etcd client during the running process. The java client I use is jetcd.

### Etcd version (please run commands below)

v3.5.4

### Etcd configuration (command line flags or environment variables)

<details>

# paste your configuration here

</details>

### Etcd debug information (please run commands blow, feel free to obfuscate the IP address or FQDN in the output)

<details>

```console
$ etcdctl member list -w table
# paste output here

$ etcdctl --endpoints=<member list> endpoint status -w table
# paste output here
```

</details>

**Comment by jbml** [comment]
Hi @luomengY,

I would suggest you provide additional details in order to facilitate the investigation of this issue. In particular, anybody who would like to look into this problem, would probably need to try and replicate it first.

Would you be able to share some more information about the program you run on the client side please ? A minimalistic java program which replicates the issue would be ideal, so that people can understand clearly what you're doing. Are you running this client program on the same machine as the etcd server ?

Also, it would be beneficial to provide some information about your etcd server setup, as requested in the template.

Finally, I feel that some insight about how/when exactly you kill the client process would help.

Thanks.

**Comment by luomengY** [comment]
Indeed my java client program is on the same server as etcd. My program is as follows:


import io.etcd.jetcd.ByteSequence;
import io.etcd.jetcd.Client;
import io.etcd.jetcd.Lease;
import io.etcd.jetcd.Lock;
import io.etcd.jetcd.lease.LeaseKeepAliveResponse;
import io.grpc.stub.StreamObserver;
import org.junit.Before;
import org.junit.Test;

import java.nio.charset.StandardCharsets;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class JEtcdTest {
    private Client client;
    private Lock lock;
    private Lease lease;
    //单位：秒
    private long lockTTl = 1;
    private ByteSequence lockKey = ByteSequence.from("/root/lock", StandardCharsets.UTF_8);
    private ScheduledExecutorService scheduledThreadPool = Executors.newScheduledThreadPool(4);
    @Before
    public void setUp() {
        client = Client.builder().endpoints(
                "http://121.4.115.5:4379"
        ).build();
        lock = client.getLockClient();
        lease = client.getLeaseClient();
    }
    @Test
    public void lockTest1toMaster() throws InterruptedException, ExecutionException {
        long leaseId = lease.grant(lockTTl).get().getID();
        lease.keepAlive(leaseId, new StreamObserver<LeaseKeepAliveResponse>() {
            @Override
            public void onNext(LeaseKeepAliveResponse value) {
                System.err.println("LeaseKeepAliveResponse value:"+ value.getTTL());
            }
            @Override
            public void onError(Throwable t) {
                scheduledThreadPool.shutdownNow();
                t.printStackTrace();
            }
            @Override
            public void onCompleted() {
                scheduledThreadPool.shutdownNow();
            }
        });
        lock.lock(lockKey, leaseId).get().getKey();
        scheduledThreadPool.submit(() -> {
            while (true) {
                System.err.println("我是主服务开始工作了");
                TimeUnit.SECONDS.sleep(1);
            }
        });
        TimeUnit.DAYS.sleep(1);
    }
    @Test
    public void lockTest2toStandby() throws InterruptedException, ExecutionException {
        long leaseId = lease.grant(lockTTl).get().getID();
        lease.keepAlive(leaseId, new StreamObserver<LeaseKeepAliveResponse>() {
            @Override
            public void onNext(LeaseKeepAliveResponse value) {
                System.err.println("LeaseKeepAliveResponse value:"+ value.getTTL());
            }
            @Override
            public void onError(Throwable t) {
                scheduledThreadPool.shutdownNow();
                t.printStackTrace();
            }
            @Override
            public void onCompleted() {
                scheduledThreadPool.shutdownNow();
            }
        });
        lock.lock(lockKey, leaseId).get().getKey();
        scheduledThreadPool.submit(() -> {
            while (true) {
                System.err.println("我是备用服务，我开始工作了，估计主服务已经挂了");
                TimeUnit.SECONDS.sleep(1);
            }
        });
        TimeUnit.DAYS.sleep(1);
    }
    @Test
    public void lockTest3toStandby() throws InterruptedException, ExecutionException {
        long leaseId = lease.grant(lockTTl).get().getID();
        lease.keepAlive(leaseId, new StreamObserver<LeaseKeepAliveResponse>() {
            @Override
            public void onNext(LeaseKeepAliveResponse value) {
                System.err.println("LeaseKeepAliveResponse value:"+ value.getTTL());
            }
            @Override
            public void onError(Throwable t) {
                scheduledThreadPool.shutdownNow();
                t.printStackTrace();
            }
            @Override
            public void onCompleted() {
                scheduledThreadPool.shutdownNow();
            }
        });
        lock.lock(lockKey, leaseId).get().getKey();
        scheduledThreadPool.submit(() -> {
            while (true) {
                System.err.println("我是备用服务，我开始工作了，估计主服务已经挂了");
                TimeUnit.SECONDS.sleep(1);
            }
        });
        TimeUnit.DAYS.sleep(1);
    }
}

**Comment by luomengY** [comment]
Please help me take a look at this problem.

**Comment by jbml** [comment]
I will try to replicate this issue on my side, @luomengY 

Just to be sure I understand what's happening in your scenario, can you confirm that you:

1. start an etcd server on a single node (with which parameters ?)
2. run the junit test cases above, which will execute lockTest1toMaster; this will create a lock in etcd with no expiration and then sleep
3. while lockTest1toMaster is sleeping, kill the junit test 

If I am not mistaken, you're explaining that at this moment, your etcd server raises a fatal error (failed to find a bucket","bucket-name":"key"), and is unable to serve any more request. Is that correct ?

**Comment by luomengY** [comment]
1.this is how I start: ./etcd --config-file etcd.conf.yml
my etcd.conf.yml:
name: hbp104
--------------------------------------------------------------------------------------
data-dir: /usr/lib/cloudmanager/components/lark/data/etcd
listen-client-urls: http://10.3.71.104:4379
advertise-client-urls: http://10.3.71.104:4379
listen-peer-urls: http://10.3.71.104:4380
initial-advertise-peer-urls: http://10.3.71.104:4380
initial-cluster: hbp104=http://10.3.71.104:4380
initial-cluster-token: etcd-cluster-hbm-token
initial-cluster-state: new
--------------------------------------------------------------------------------------------
2.lockTest1toMaster(), lockTest2toStandby() and lockTest3toStandby() are for master/standby switching. lockTest1toMaster(), lockTest2toStandby() and lockTest3toStandby() will start at the same time, continuously starting and stopping one or both of them.
My java client program and etcd are placed on the same server. I found that when I run the java client program locally, remote access etcd doesn't seem to have this problem.
3.The error is: "failed to find a bucket", but sometimes there is an error: "range failed to find revision pair".

**Comment by luomengY** [comment]
1.this is how I start: ./etcd --config-file etcd.conf.yml
my etcd.conf.yml:

--------------------------------------------------------------------------------------------------
name: hbp104
data-dir: /usr/lib/cloudmanager/components/lark/data/etcd
listen-client-urls: http://10.3.71.104:4379/
advertise-client-urls: http://10.3.71.104:4379/
listen-peer-urls: http://10.3.71.104:4380/
initial-advertise-peer-urls: http://10.3.71.104:4380/
initial-cluster: hbp104=http://10.3.71.104:4380/
initial-cluster-token: etcd-cluster-hbm-token
initial-cluster-state: new
--------------------------------------------------------------------------------------------------------------
2.lockTest1toMaster(), lockTest2toStandby() and lockTest3toStandby() are for master/standby switching. lockTest1toMaster(), lockTest2toStandby() and lockTest3toStandby() will start at the same time, continuously starting and stopping one or both of them.
My java client program and etcd are placed on the same server. I found that when I run the java client program locally, remote access etcd doesn't seem to have this problem.
3.The error is: "failed to find a bucket", but sometimes there is an error: "range failed to find revision pair".

**Comment by luomengY** [comment]
1.this is how I start: ./etcd --config-file etcd.conf.yml
my etcd.conf.yml as follow:
-----------------------------------------------------------------------------------------------------------------
name: hbp104
data-dir: /usr/lib/cloudmanager/components/lark/data/etcd
listen-client-urls: http://10.3.71.104:4379
advertise-client-urls: http://10.3.71.104:4379
listen-peer-urls: http://10.3.71.104:4380
initial-advertise-peer-urls: http://10.3.71.104:4380
initial-cluster: hbp104=http://10.3.71.104:4380
initial-cluster-token: etcd-cluster-hbm-token
initial-cluster-state: new
-------------------------------------------------------------------------------------------------------------------
2.lockTest1toMaster(), lockTest2toStandby() and lockTest3toStandby() are for master/standby switching. lockTest1toMaster(), lockTest2toStandby() and lockTest3toStandby() will start at the same time, continuously starting and stopping one or both of them.
My java client program and etcd are placed on the same server. I found that when I run the java client program locally, remote access etcd doesn't seem to have this problem.
3.The error is: "failed to find a bucket", but sometimes there is an error: "range failed to find revision pair".

**Comment by luomengY** [comment]
I did another experiment. When my java client runs on a server outside the etcd cluster, this problem does not occur. When the java client and etcd are on the same server, is there any special attention?

**Comment by luomengY** [comment]
And I found that when my java client program is on the same server as etcd, every time I stop and restart the java client, it will report an error: "range failed to find revision pair".

**Comment by jbml** [comment]
I have not been able to replicate this issue on my side. 

Here are the steps I followed:
 - start an etcd server on a linux box with one node, using release 3.5
 - start lockTest1toMaster: the logs show that this one has the lock
 - start lockTest2toStandby: the logs show that this one is waiting for the lock
 - start lockTest3toStandby: the logs show that this one is waiting for the lock
 
Then I interrupt lockTest1toMaster, and the logs show that lockTest2toStandby gets the lock. 
Then I interrupt lockTest2toStandby, and the logs show that lockTest3toStandby gets the lock. 

I don't get the same errors that you had, and my etcd server is still functional (able to put and get keys) after these actions.

Also I don't see why the server would behave in a different way, whether the client runs on the same box or not,

**Comment by luomengY** [comment]
Is your client java program and etcd on the same server? This error occurs when etcd and the client program are running on the same server.

**Comment by jbml** [comment]
Yes, this was all on the same server.

**Comment by luomengY** [comment]
ok, I have found the reason, due to the inconsistency of the metadata of my etcd after starting and stopping.

**Comment by jbml** [comment]
Do you mind explaining a bit more your findings @luomengY please, so that people facing the same error get some clues about it  ?

Is there anything to investigate in regards to the metadata inconsistency you found ?

**Comment by luomengY** [comment]
The problem I encountered is because our platform has a master-slave switchover, and there is a master-slave synchronization data directory. We also put the etcd data in this directory, which is equivalent to manually synchronizing the data, resulting in once the master-slave switchover , which may cause changes to the primary and secondary metadata.

**Comment by jbml** [comment]
Thanks for the update @luomengY 

I can indeed see how this issue is related to your particular setup, and not etcd itself.
I would suggest we close this issue then, as no further action is required.

**Comment by luomengY** [comment]
ok

---Original---
From: "Jeremy ***@***.***&gt;
Date: Sat, Aug 13, 2022 21:17 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [etcd-io/etcd] When I killed my java etcd program, I found thatetcd would stop working and report the following error. What is the reason forthis? (Issue #14314)




 
Thanks for the update @luomengY
 
I can indeed see how this issue is related to your particular setup, and not etcd itself.
 I would suggest we close this issue then, as no further action is required.

**Comment by ahrtr** [comment]
Close this issue per the discussion

**Comment by luomengY** [comment]
Is your client java program and etcd on the same server? This error occurs when etcd and the client program are running on the same server.


---Original---
From: "Jeremy ***@***.***&gt;
Date: Tue, Aug 9, 2022 15:34 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [etcd-io/etcd] When I killed my java etcd program, I found thatetcd would stop working and report the following error. What is the reason forthis? (Issue #14314)




 
I have not been able to replicate this issue on my side.
 
Here are the steps I followed:
  
start an etcd server on a linux box with one node, using release 3.5 

start lockTest1toMaster: the logs show that this one has the lock 

start lockTest2toStandby: the logs show that this one is waiting for the lock 

start lockTest3toStandby: the logs show that this one is waiting for the lock 
 
Then I interrupt lockTest1toMaster, and the logs show that lockTest2toStandby gets the lock.
 Then I interrupt lockTest2toStandby, and the logs show that lockTest3toStandby gets the lock.
 
I don't get the same errors that you had, and my etcd server is still functional (able to put and get keys) after these actions.
 
Also I don't see why the server would behave in a different way, whether the client runs on the same box or not,
